### ISSUE TITLE
Increasing CPU memory during training stage

### ISSUE THREAD

**Issue by dalision** [espnet/espnet#4632]
Hi, maintainers! When I run an ASR trainer script, I observed an increasing CPU memory each step by psutil.virtual_memory()(about 2-3MB), while GPU memory is relatively stable. I have noted the code about Reporter ( it may cause the problem because it will accumulate process variables). And I check some other variables like loss(loss.detach() are used). However, the problem still exists. OOM or dataloader worker pipe broken may happen if the steps number in an epoch is big enough. I wonder what the reason for that. Thanks!

**Comment by sw005320** [comment]
Thanks for the report.
Does "each step" mean "each iteration"?
If so, this would be a critical issue.

Could you tell me more detailed information? (e.g., PyTorch versions etc. based on our bug issue template?)

**Comment by dalision** [comment]
Hi,sorry for late reply. Yes, "each step" mean "each iteration". Here are more details.

OS:CentOS Linux release 7.2.1511(Core) 
others:python 3.7.10+torch 1.71+cuda10.2+espnet 0.10.7a1
device: Teslav100-PCIE-12GB
We trained a 12-layer-conformer on misp2021 dataset on ASR Task.



![图片](https://user-images.githubusercontent.com/46208723/189832827-07afcfdc-8b7e-488b-82b8-2713cdaa00b5.png)



I add the following code in the function train_one_epoch of espnet2.train.trainer.py to observe the CPU and GPU memory change during the training stage.

```
@classmethod
import psutil
def get_avaliable_memory(cls):
	meminfo = torch.cuda.memory_allocated()
	logging.info("GPU Allocate memory"+":"+str(round(meminfo/1024**2+"MB")))
	logging.info("GPU Reserve memory"+":"+str(round(meminfo/1024**2+"MB")))
	mem = psutil.virtual_memory()
	logging.info("cpu available memory"+":"+ str(round(mem.used/1024**2))"+'MB')
```

I observed that in each iteration CPU memory will continuously increase (about 2-5MB each iteration). As I mentioned above I ignore some variate(like Reporter) that may cause this problem.

**Comment by kamo-naoyuki** [comment]
Reporter maintains all registered values for an epoch. This causes increasing memories for every iteration.

I don't have any ideas for now to solve the issue.

**Comment by dalision** [comment]
I konw Reporter can cause such problem, so I delete the ralated code with Reporter like reporter.register or reporter.next as I have mentioned above. But cpu memory still increase. It is easy to reproduce if this problem does exists. You may have a try.

**Comment by kamo-naoyuki** [comment]
Okay, hmm, I'll investigate it.

**Comment by kamo-naoyuki** [comment]
@dalision I couldn't reproduce the issue.

Did you set `num_workers>=1`(The default is 1)?  `num_workers>=1 ` causes increasing memory. The pytorch DataLoader launches worker processed for creating mini-batches, and they are cached until sending it to the main process. If the speed of creating mini-batches is too faster than the speed of a forward-backward process,  it cause broken pipe. This is not a bug.


If you met the issue even if you set `num_workers==0`, to clarity which process causes memory leaks, please also check `psutil.Process(os.getpid()).memory_info()` instead of `psutil. virtual_memory` .

**Comment by dalision** [comment]
Thanks for your early reply. I guess you are right. Actually I first find this problem based on a modified trainer.py, I did set workernum==0. And then I reproduced the problem on the original trainer.py but set num_worker>0. I will take your advice for further investigation. So I temporarily close the issue.
